pylib/tools/hookslib.py

```diff
@@ -121,15 +129,26 @@ def __call__({", ".join(args_including_self)}) -> None:
                 self._hooks.remove(hook)
                 raise
 """
-        if self.legacy_hook:
+        if self.replaces:
+            out += f"""\
+        if {self.replaces}.count() > 0:
+            print(
+                "The hook {self.replaces} is deprecated.\\n"
+                "Use {self.name} instead."
+            )
+            {self.replaced_args()}
+"""
+        elif self.legacy_hook:
+            # don't run legacy hook if replaced hook exists
+            # otherwise the legacy hook will be run twice
```

Comment:
The replacing new-style hook invokes the previous new-style hook and the previous new-style hook invokes the legacy hook. If the replacing new-style hook would invoke the legacy hook again, the legacy hook would be run a second time.

So the if ... elif here is correct. But I noticed another error: The previous new-style hook should always be invoked by the replacing one, not only `if {self.replaces}.count() > 0`.

Reply:
> If I add it to addon_config_editor_will_update_json instead, ...

My understanding of the word "legacy" in the term "legacy hook" is that they should not be used for new functionality. The way I understand it, we only need to ensure that existing legacy hooks are invoked, which happens by invoking the previous new-style hook.

Reply:
To be fair, following the reasoning that legacy hooks shold not be used for new hook definitions, there should never be a case where both `self.replaces` and `self.legacy_hook` are not `None`.
Should I add a check for this in hookslib and raise an error to catch faulty hook definitions?
Something like
```python
if self.replaces and self.legacy_hook:
    raise SomeError
```

Reply:
You're right that the hook shouldn't be added to new code, but I was thinking of a case where a developer may move the legacy hook from one hook to another. It's admittedly a corner case and may never happen, but I'd feel a bit better if the code raised an error like you suggested (eg raise Exception("can't use both replaces and legacy_hook")) instead of just silently ignoring one of the provided inputs. Sorry to be picky here :-)

Reply:
No worries.
I added a line about not defining legacy hooks in new hooks in the docs as well.